### DISCUSSION TITLE
Madimack Elite V3 pool heatpump Fault Code are a bitmap field, how to resolve

### DISCUSSION THREAD

**Discussion by sikkek** [make-all/tuya-local#2154] [Q&A] [answered]
Hi all,

I'm working on improving the Madimack Elite V3 pool heatpump definition. According to Query Things Data Model on the Tuya Developer platform, regarding fault codes, the heat pump returns the following 2 dpids:
`{
	"abilityId": 15,
	"accessMode": "ro",
	"code": "fault",
	"description": "",
	"extensions": {
		"scope": "fault",
		"attribute": "128"
	},
	"name": "故障告警",
	"typeSpec": {
		"type": "bitmap",
		"label": [
			"E1",
			"E2",
			"E3",
			"E4",
			"E5",
			"E6",
			"E7",
			"E8",
			"E9",
			"EA",
			"EB",
			"ED",
			"P0",
			"P1",
			"P2",
			"P3",
			"P4",
			"P5",
			"P6",
			"P7",
			"P8",
			"P9",
			"PA",
			"F1",
			"F2",
			"F3",
			"F4",
			"F5",
			"F6",
			"F7"
		],
		"maxlen": 30
	}
}`
and

`{
	"abilityId": 101,
	"accessMode": "ro",
	"code": "fault1",
	"description": "",
	"extensions": {
		"scope": "fault"
	},
	"name": "故障1",
	"typeSpec": {
		"type": "bitmap",
		"label": [
			"F8",
			"F9",
			"Fb",
			"Fa"
		],
		"maxlen": 4
	}
}`

Both data points seem to be bit flags, with more then 1 possible bit set. BTW: I did confirm that on dpid 15, value 4 (binary 0000 0100) equals error E3. (this is a no water flow error which I can easily trigger)

So: how can I alter the device definition that correctly translates these bit flags to Home Assistant? Writing out all possible bit combinations is not feasible (2^30 is roughly 1 billion).

**Comment by make-all** [comment, marked as answer]
The type `bitfield` is designed to handle this (I came up with the type name before I saw that Tuya uses bitmap, and I avoided that name to avoid confusion with the image type).

The difference between `integer` and `bitfield` is that when comparing mappings, instead of a normal = comparison, the bitfield compares only the bits that are set (current_value & mapping_value = mapping_value).

So your mapping just needs to match the individual values you want to match.

Since there can only be a single value at the end, the first match is used, so if you want to detect multiple values at once, use a binary_sensor for every important thing you want to detect.

**Reply by sikkek** [reply]
Makes sense. However, this means that in the case of this heatpump, I need to create 34 binary_sensors, which used in a dashboard requires a lot of space. Wouldn't it be possible to create a helper that, given the input value and the mapping array, return a concatenated string with the 'ON' codes?

**Reply by make-all** [reply]
It is unlikely that all of those fault codes are useful.  Normally a few important or frequently occurring ones are pulled out, and the rest lumped under a general problem binary_sensor.  Or you could group them into compressor related, pump related, etc to reduce the binary sensors needed.
